students/scott_bromley/lesson03/mailroom.py

```diff
@@ -0,0 +1,142 @@
+
+
+def send_thank_you():
+    '''
+    prompts user for a donor name; if name exists, prompts for donation, otherwise adds to donors
+    list displays a list of donors
+    generates
+    :return: None
+    '''
+    full_name = get_user_input("Please enter the name of a donor: > ").title()
+    for donor in donors:
+        if full_name in donors:
+            donation = float(get_user_input("Please enter the donation amount: $").strip('$'))
+            add_donation(full_name, donation)
+            format_thank_you(full_name, donation)
+            break
+        elif full_name.lower() == "list":
+            print(f"List of donors: {', '.join(get_donors(donors))}")
+            break
+        else:
+            print(f"Sorry, the name you entered is not in our donor list.  Adding donor to list...")
+            donors.append((full_name, []))
+            donation = float(get_user_input("Please enter the donation amount: $").strip('$'))
+            add_donation(full_name, donation)
+            print(format_thank_you(full_name, donation))
+            break
+    return None
+
+
+def format_thank_you(full_name, donation):
+    """
+    :param full_name: name of the donor
+    :param donation: the amount donated
+    :return: None
+    """
+    now = datetime.datetime.now()
+    thank_you = (
+        f"\n\n{now.date()}\n\n"
+        f""
+        f"Dear {full_name},\n\n"
+        f""
+        f"On behalf of all of us at the ADL office we'd like to recognize and thank you for your generous\n"
+        f"donation of ${donation}.  These funds will be used to further our mission in fighting hate\n"
+        f"through leadership programs and education.  \n\n"
+        f"We look forward to seeing you at this year's banquet.\n"
+        f"\n"
+        f"Sincerely, "
+        f"\n"
+        f"ADL")
+    return thank_you
+
+
+def create_report():
+    """
+    prints a formatted table of donors ranked in descending order by Total Given
+    :return: None
+    """
+    names, totals, num_gifts, avg_gift = get_donor_summary(donors)
+    print(f"Donor Name{'':<20} | Total Given{'':>0} | Num Gifts{'':>0} | Average Gift{'':>0}")
+    print(f"-" * 72)
+    for name, total, num_gift, avg_gift in zip(names, totals, num_gifts, avg_gift):
+        print(f"{name:<32}${total:>11}{num_gift:>12}  ${avg_gift:>13}")
+    return None
+
+
+def get_donor_summary(donors=None):
+    """
+    :param donors: optional list of donors and donations
+    :return: lists of names, totals, number of gifts, avg gift sorted by highest total
+    """
+    sorted_donors = sorted(donors, key=lambda x: sum(x[1]), reverse=True)
+    names = tuple(map(lambda x: x[0], sorted_donors))
+    totals = tuple(map(lambda x: float(format(sum(x[1]), '.2f')), sorted_donors))
+    num_gifts = tuple(map(lambda x: len(x[1]), sorted_donors))
+    avg_gift = tuple(map(lambda x: float(format(statistics.mean((x[1])), '.2f')), sorted_donors))
+    return names, totals, num_gifts, avg_gift
```

Comment:
above is actually pretty inefficient - there are a total of 4 loops when all of this can be done in one loop and put into a list structure which you create later anyway.